ARCs/arc-0060.md

```diff
@@ -0,0 +1,449 @@
+}
+```
+
+- `data` must be validated with the JSON schema in `metadata` (defined below).
+- `signer` is a `Ed25519Pk` object (defined below) that represents the signer public key.
+
+#### Interface `Signature`
+
+`Signature` is the produced 64-byte array, ed25519 digital signature of the signed message.
+
+```tsx
+export type Signature = Uint8Array;
+```
+
+#### Enum `ScopeType`
+
+The `ScopeType` enumerates constant strings with the scope of a signing action.
+
+This ARC introduces two scope types.
+
+| ScopeType | Description |
+| --- | --- |
+| AUTH | Signature of a basic 32 bytes challenge for authentication. |
+| LSIG | Signature of an Algorand program for delegation. |
+
+For each `ScopeType` there is associated one JSON schema (see below).
+
+Any extension of this ARC **SHOULD** adopt an existing `ScopeType`, or introduce a new one (with the respective JSON schema).
+
+#### Interface `StdSignMetadata`
+
+A `StdSignMetadata` object specifies the metadata for the `StdSigData`.
+
+```tsx
+export interface StdSignMetadata {
+    /**
+    * The scope value of the signing data request.
+    */
+    scope: ScopeType;
+
+    /**
+    * JSON schema for the signing data.
+    */
+    schema: string;
+
+    /**
+    * Optional encoding used to represent the signing bytes in the schema.
+    */
+    encoding?: string;
+}
+```
+
+- `schema` JSON schema to be used according to the `scope`. ARC-60 compliant JSON schemas must have two mandatory fields:
+  - `ARC60Domain` domain separator for the signing bytes.
+  - `bytes` signing bytes.
+- `encoding` specifies the encoding used for the `schema.bytes` field. If not specified, encoding should be `base64`.
+
+> The JSON schema is inspired by <a href="https://eips.ethereum.org/EIPS/eip-712">EIP-712: Typed structured data hashing and signing proposal</a>.
+
+##### Example: AUTH JSON Schema
+
+JSON schema for signing a 32-byte array representing an authentication challenge; `ScopeType=AUTH`. In this example, the `StdDataStr` **MUST** adhere to the following schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "ARC60Domain": {
+      "type": "string",
+      "description": "The ARC-60 domain separator"
+    },
+    "bytes": {
+      "type": "array",
+      "items": {
+        "type": "integer",
+        "minimum": 0,
+        "maximum": 255
+      },
+      "description": "Challenge to be signed for authentication",
+      "maxItems": 32
+    }
+  },
+  "required": ["ARC60Domain", "bytes"]
+}
+```
+
+- The `ARC60Domain` object indicates the domain separator to be used for signing. It **SHOULD** be set to `"arc60"` or empty `""`. Algorand transaction domain separators `TX` and `TG` **MUST** be forbidden.
+- The `bytes` object is the byte array of data being signed. In the example, it is a 32-byte challenge to be signed for authentication scope. The `bytes` field cannot be a valid Algorand transaction, therefore it **MUST NOT** be prepended with a known domain separator `TX` or `TG`.
```

Comment:
Is ruling out certain prefixes in the signing `bytes` a good idea? 

Assume an application creates a random 32-byte string that needs to be signed as a challenge. If that byte string starts with `TX`, `TG`, or any other forbidden prefix, the wallet must reject it.  This may be confusing to an (unaware) application: most of the time the wallet is accepting its challenge but sometimes it is refusing to sign the data because the random data happens to have a forbidden prefix.

Alternatively, what if we force `ARC60Domain` to be non-empty and to not be one of a set of blacklisted prefixes (`TX`, `TG`, ...)? Then we wouldn't have to restrict the `bytes`.

Reply:
1. `Program` should be forbidden only **IF** `ScopeType` != `LSIG`. I am clarifying this in the semantics as well. And yes, `TG` is for groups https://github.com/algorand/go-algorand/blob/981abad60d1fdd2993847434417982b4d373b7b8/protocol/hash.go#L74 
2. This problem is difficult to resolve and maybe it is a UX/UI issue. Auth protocols working with a `challenge-response` have no idea of our prefix-thing. Your proposal would break the compatibility with those protocols and I don't like it. For example see [here](https://www.w3.org/TR/webauthn-2/#sctn-sample-authentication)